flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSource.java

```diff
@@ -110,69 +106,64 @@ public Boundedness getBoundedness() {
-        // sources.forEach(t -> serializers.add(castSerializer(t.source.getSplitSerializer())));
         return new HybridSourceSplitSerializer(switchedSources);
     }
 
     @Override
     public SimpleVersionedSerializer<HybridSourceEnumeratorState>
             getEnumeratorCheckpointSerializer() {
-        List<SimpleVersionedSerializer<Object>> serializers = new ArrayList<>();
-        sources.forEach(
-                t -> serializers.add(castSerializer(t.source.getEnumeratorCheckpointSerializer())));
-        return new HybridSourceEnumeratorStateSerializer(serializers);
-    }
-
-    private static <T> SimpleVersionedSerializer<T> castSerializer(
-            SimpleVersionedSerializer<? extends T> s) {
-        @SuppressWarnings("rawtypes")
-        SimpleVersionedSerializer s1 = s;
-        return s1;
+        return new HybridSourceEnumeratorStateSerializer(switchedSources);
     }
 
     /**
-     * Callback for switch time customization of the underlying source, typically to dynamically set
-     * a start position from previous enumerator end state.
+     * Factory for underlying sources of {@link HybridSource}.
      *
-     * <p>Requires the ability to augment the existing source (or clone and modify). Provides the
-     * flexibility to set start position in any way a source allows, in a source specific way.
-     * Future convenience could be built on top of it, for example an implementation recognizes
-     * optional interfaces.
+     * <p>This factory permits building of a source at graph construction time or deferred at switch
+     * time. Provides the ability to set a start position in any way a specific source allows.
+     * Future convenience could be built on top of it, for example a default implementation that
+     * recognizes optional interfaces to transfer position in a common format.
      *
      * <p>Called when the current enumerator has finished and before the next enumerator is created.
-     * The enumerator end state can thus be used to set the next source's start start position.
+     * The enumerator end state can thus be used to set the next source's start start position. Only
+     * required for dynamic position transfer at time of switching.
      *
-     * <p>Only required for dynamic position transfer at time of switching, otherwise source can be
-     * preconfigured with a start position during job submission.
+     * <p>If start position is known at jib submission, the source can be constructed in the entry
+     * point and simply wrapped into the factory, providing the benefit of validation during
+     * submission.
      */
-    public interface SourceConfigurer<SourceT extends Source, FromEnumT extends SplitEnumerator>
+    public interface SourceFactory<T, SourceT extends Source, FromEnumT extends SplitEnumerator>
             extends Serializable {
-        SourceT configure(SourceT source, FromEnumT enumerator);
+        SourceT create(FromEnumT enumerator);
```

Comment:
Enumerators are currently more or less implementation details. We might need to add another redirection to make it sound. We already had the idea of a universal enumerator state. So would change the method to 

```
SourceT create(CommonEnumState enumeratorState, CreationContext context);
```

Then either the enumerator implements some interface to generate that state. Or the user would have to supply such an extractor (which effectively bypasses encapsulation but gives the opportunity to use sources without the respective enumerator).

I'm sure there are also other options.

Reply:
On a second thought, having that interface is a bit lackluster. 
I think I'd just expose everything that we have and let the user pick.
```
SourceT create(SourceSwitchContext<...> context);

class SourceSwitchContext<...> {
  int index;
  SourceX lastSource;
  EnumT enumerator;
  CheckpointT finalEnumState;
  List<SplitStateT> finalReaderStates;
  //
  getters;
}
```
and hope that some information is usable. I especially hope that the states actually provide exactly what we need.

Reply:
I am actually a little confused by this API. How would the next source be constructed from the previous `SplitEnumerator`? Do we expect the users to understand the information contained in the `EnumCheckpointT` of a `SplitEnumerator`? If so, I feel that such hidden protocol is quite fragile as by design the `EnumCheckpointT` is only generated and parsed by the same `SplitEnumerator`, i.e. purely internal state and subject to change at any time. Depending on such internal state seems a little hacky.

Reply:
The next source can be constructed from any information the specific enumerator provides, such as an end timestamp. In case you refer to the context suggested above, additional information may be provided, but I have not looked into that yet. See example for API usage in javadoc: https://github.com/apache/flink/pull/15924/files#diff-fe96bc815d8413a73df83b0d034a98184646a2ab798be6c9774203c7919cdf88R44 (tests may also be helpful)

Reply:
At a high level, in terms of how to perform the switchover, I think we need to answer the following questions:
1. **How to extract the _end_position_ information from the previous source?** I agree it is intuitive to get the _end_position_ from `SplitEnumerator` of the previous source. However, it seems some protocol is missing here. We probably need to define some method like `getEndState()` in the `SplitEnumerator` interface so users knows how to get the end state, no matter what the end state is.
2. **How to translate the _end_position_ of the previous source to the _start_position_ of the next source?** This could be done in the `SourceFactory`. I have no concern about this.
3. **How to construct the next source from the _start_position_?** This is also done in the `SourceFactory`. I agree having the `SourceFactory` here makes sense as it allows users to build the next Source dynamically.

My concern here is just for the first question.
I am wondering how would the `SourceFactory` for the next source get the information (timestamp, etc) from the enumerator of the previous source? The `SplitEnumerator` interface does not have a method to get such information. Does that mean the users need to use the concrete enumerator classes (e.g. `KafkaPartitionSplitEnumerator`) and these concrete classes also required to expose the information for switching via some additional method?

For the example in the java doc, what would be an expected implementation of `createKafkaSource`? It seems a little difficult to derive the start position of the Kafka source from a `StaticFileSourceEnumerator`, unless it is modified by the users for this specific use case.

Reply:
@AHeise I implemented the `SourceSwitchContext`. It only contains the previous enumerator as that is currently the only component that I can see provides useful contextual information. Now that we have this extension point we can add more in the future; once scenarios that would benefit from it become clear. 

Note that in order to make additions to the context backward compatible, we cannot rely on additional context type parameters though (limited to EnumT). I think that's fine since earlier in this PR the daunting list of type parameters was already identified as a drawback and they provide minimal benefit in this case.

Reply:
@becketqin thank you for the detailed analysis. I agree that there should be a way to extract the end position from `FileSource`. It's mentioned on the FLIP that we would like to keep this work separate, so that we can also think about a potential uniform end position format. Meanwhile `SourceFactory` and `SourceSwitchContext` provide the mechanism to solve this in a source specific way. We intend to use it internally with a custom source and other sources may provide their own way to access the end position. I hope that the augmented examples make this more clear and the intended boundary for this PR also makes sense? CC: @AHeise

Reply:
@tweise Thanks for the explanation and updating the PR. So it looks that the overall design here is following:

1. The `HybridSource` manages the process of switching between two sources.
2. A user provided implementation of `SourceFactory` is used to create the next `Source` when the previous `Source` finishes reading.
3. The `SourceFactory` is expected to do the following:
     a. Get the _end_position_ from the `SplitEnumerator` of the previous finished `Source`.
     b. Translate that _end_position_ to the _start_position_ of the next source.
     c. Construct and setup the next Source.

Regarding 3.a, users should not expect the existing `Source` implementations to work with the `HybridSource` out of the box. Users are expected to create their own `Source` whose `SplitEnumerator` exposes the _end_position_ after the source finishes reading.

There is a plan to submit a separate FLIP to discuss the API which unifies the way the `SplitEnumerator`s should expose the _end_position_. 


To be honest, I am still a little concerned about putting the _end_position_ retrieval API into a separate FLIP. If we do that, users would have to first implement a `Source` of their own way to expose the end position. And later on when we add the _end_position_ retrieval API, users likely have to change their code again to adopt that API. In that case, we may be introducing something that is known to be changed backwards incompatibly in the future.

Please correct me if I misunderstand something.

Reply:
@becketqin your description of the switching process is correct but it only covers the scenario where the start position of the subsequent source is determined at switch time. There is the much simpler scenario with fixed start position which does not require the user to implement `SourceFactory`:

```
        FileSource<String> fileSource =
                FileSource.forRecordStreamFormat(new TextLineFormat(), Path.fromLocalFile(testDir)).build();
        KafkaSource<String> kafkaSource =
                KafkaSource.<String>builder()
                        .setBootstrapServers("localhost:9092")
                        .setGroupId("MyGroup")
                        .setTopics(Arrays.asList("quickstart-events"))
                        .setDeserializer(
                                KafkaRecordDeserializer.valueOnly(StringDeserializer.class))
                        .setStartingOffsets(OffsetsInitializer.earliest())
                        .build();
        HybridSource hybridSource =
                HybridSource.builder(fileSource)
                        .addSource(kafkaSource)
                        .build();

```

The intention for this PR is to cover that simple scenario and at the same time provide the mechanism to build something more advanced with a different bounded source. This isn't in conflict with subsequent addition of standard interface for extracting end position and we can discuss whether or not that should be part of FLIP-150. However I hope to find a way to move this PR forward independently as it has been under review for a long time now.

Reply:
@tweise Thanks for the explanation. That makes sense to me. 

Actually after a second thought, maybe we don't need a unified API on the _end_position_.

Such _end_position_ may vary in different organizations and use cases even for the same Source. For example, a Source reading from HDFS may either define the _end_position_ as a timestamp, or a list of Kafka TopicPartition-offset pairs indicating the exact messages put into the file when the file was closed. The the definition of _end_position_ largely depends on individual users. It is difficult to provide a reasonable implementation of _end_position_ out of the box. Arguably, timestamp could be an option applicable to all the Sources, but there are still a lot of other cases. Therefore, it seems inevitable for the users to implement their own way to define and expose the _end_position_. That means the _end_position_ retrieval has to be `SplitEnumerator` implementation specific. So having a unified _end_position_ retrieval API in the `SplitEnumerator` does not have a strong use case support.

It probably has been clear to everyone else except me... Sorry for the long discussion and thanks for the patient explanation. The current API looks good to me. I have no further concerns now.